### PR TITLE
Refactoring to allow new Maps GL engine

### PR DESCRIPTION
**Pull request by turban**
This PR will align the Leaflet API with the new Maps GL API, making it easier to switch between the two. 

- [x] Dependency upgrades.
- [x] New folder structure (layers and controls in separate folders).
- [x] Map is not extending L.Map to restrict available methods.
- [x] Unify event callback response.
- [x] Unify how we pass coordinates around - we now use  [lng, lat] as in GeoJSON. 
- [x] New "hasLayerSupport" method on Map to check if a certain layer type is supported. We use this to exclude Google layers when using WebGL. 
- [x] New "openPopup" method on Map, to open a popup on a coordinate location.
- [x] New mixin with shared layer methods. Will be implemented differently in maps-gl where all layers extend the same Layer object. 
- [x] New setVisibility method to toggle layer visibility.
- [x] New "onClick" and "onRightClick" layer options to unify event response. 
- [x] New "setIndex" method to control layer ordering.
- [x] New "orderLayers" method on Map to sort layers according to layer index.
- [x] Create layer pane when layer is added
- [x] Group labels and buffers with associated layers (not as separate layers)
 
Additional unit tests will not be added to this repo as it soon will be replaced by maps-gl (which should have good test coverage).

### CODE REVIEW

**Comment by amcgee** [comment]
Sorry for the delay in reviewing this @turban !

In testing this locally I've found a number of layer types which cause top-level Javascript errors.  Specifically:

* Opening any Boundary layer
* Creating a new boundary layer
* Opening or re-creating the "Malaria: Cases Sierra Leone 2015-2016 clustered" map.

See screenshots below - there may also be others I haven't found.  The setup I used to test this was:

* Checkout the `maps-gl` branch of both `gis-api` and `maps-app`
* Comment out the `path.resolve(__dirname, '../gis-api/src/')` line in webpack config for maps-app (since my `gis-api` is located somewhere else, and also so I could test the built version of `gis-api`)
* Run `yarn install` and `yarn build` in `gis-api`
* `yarn link` from `./build` in `gis-api`
* `yarn link @dhis2/gis-api` in `maps-app`
* `Update all includes in `maps-app` from `gis-api/src` to just `gis-api`, since I'm using the built version.

screen shot 2019-02-19 at 11 20 50 pm

screen shot 2019-02-19 at 11 20 34 pm

screen shot 2019-02-19 at 11 21 09 pm

*The above error appeared silently (no layer added) when creating a new Boundary layer*

screen shot 2019-02-19 at 11 22 33 pm

File: src/layers/GeoJson.js [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+    onAdd(map) {
+        const {
+            opacity,
+            isVisible,
+            onClick,
+            onRightClick,
+            highlightStyle,
+        } = this.options;
+
+        L.GeoJSON.prototype.onAdd.call(this, map);
```
Comment (amcgee) [inline src/layers/GeoJson.js]: Why do we need to use `.call` for these methods?

**Review by amcgee — CHANGES_REQUESTED** [review]
Didn't look too deeply into the code yet, just ran some tests locally

File: src/layers/GeoJson.js (hunk 1 above)
Comment (turban) [reply src/layers/GeoJson.js]: See: https://leafletjs.com/examples/extending/extending-1-classes.html#methods-of-the-parent-class

**Comment by turban** [comment]
@amcgee, thanks for reviewing - I've pushed a fix for the boundary layer issue.

**Comment by amcgee** [comment]
@turban boundary layers look good!

There is still the issue with loading `Malaria: Cases Sierra Leone 2015-2016 clustered` map.  The top-level error I'm seeing is `TypeError: bounds.map is not a function`

screen shot 2019-03-01 at 5 53 48 am
screen shot 2019-03-01 at 5 54 07 am

**Comment by amcgee** [comment]
I get a different error `Cannot convert undefined or null to object` when trying to create the same layer from scratch (Malaria EventLayer with 01/01/2015-12/31/2016 period):

screen shot 2019-03-01 at 5 58 04 am

**Comment by turban** [comment]
Good catch! I'm pushed a fix which will hopefully sole both issues with server clusters.

**Review by amcgee — APPROVED** [review]

**Comment by amcgee** [comment]
@turban LGTM!!  👍

**Pull request merged**